131 add solana support

notorious-d-e-v opened the pull request:
# Add Solana Virtual Machine (SVM) Support to x402 Protocol

## Overview

This pull request adds Solana support for X402 according to the discussion in https://github.com/coinbase/x402/issues/131 and the final [sequence diagram](https://github.com/coinbase/x402/issues/131#issuecomment-3059196377) proposed by @ngundotra.

The implementation includes an `exact` payment scheme for Solana.

## Key Features

### 🚀 Core SVM Implementation
- **Exact Payment Scheme for Solana**: Complete implementation of the `exact` payment scheme for Solana Virtual Machine
- **SVM Client Support**: Client-side utilities for creating and signing Solana transactions
- **SVM Facilitator**: Payment verification and settlement
- **Transaction Processing**: Full support for partially-signed transactions with fee payer sponsorship by facilitator

### 📦 Package Support
 - x402-fetch
 - x402-express
 - x402-axios (todo after feedback)
 - x402-hono (todo after feedback)
 - x402-next (todo after feedback)

### 📄 Examples Support
 - clients/fetch
 - facilitators/svm/express
 - servers/express (todo after feedback)
 - clients/axios (todo after feedback)

### 🧪 Testing & Quality
- **Comprehensive Test Suite**: Unit tests for all SVM components including verification, settlement, and fee payer functionality
- **Error Handling**: Robust error handling and validation throughout the SVM implementation

## Technical Implementation Details

### Fee Payer

In order for the facilitator to sponsor a transaction, the client needs to know the address of the facilitator's signer that will sign the transaction and pay the fee.

**Current Implementation**: Added a `/fee-payer` endpoint to the facilitator
- Made it a POST request to send `paymentRequirements` as part of the JSON body
- This allows the facilitator to return different fee payer addresses based on scheme and network in the future if necessary
- **Considerations**:
  - Could rename to `/hints` for broader use cases
  - Could simplify to GET request with query parameters for network and scheme
  - Could integrate into the `/verify` endpoint to reduce network calls
  - Benefits of `/verify` integration:
    - Future schemes may need additional hints (priority fees, blockhash, etc.)
    - Simplifies server middleware code -- only proxy the request to the facilitator verify/settle

### NetworkEnum Type Addition

Introduced a `NetworkEnum` type to standardize network identifiers across the codebase.

String based networks are more aesthetic but may lead to frustrating typo errors.

I may have gone trigger happy with this addition and am happy to remove it if it adds unnecessary complexity.
It seems to touch a lot of code and may break something unintentionally.
The string-based approach was working fine, and this could be considered out of scope.

### Network Calls

#### Client-Side Network Calls
3 calls total while creating payment header:
- **Blockhash Retrieval**: RPC call to get blockhash for transaction expiry (validUntil) time
- **Token Program (smart contract) Detection**: RPC call to determine if asset belongs to Token or Token-2022 program
- **Asset Decimals**: RPC call to get asset decimals for transfer instruction creation

#### Facilitator-Side Network Calls
4 network calls to Solana RPC during verification
- **Verify Endpoint**:
  - Lookup tables fetch for transaction decompilation during introspection
  - Associated Token Account (ATA) fetch for receiver during transaction introspection
  - Associated Token Account (ATA) fetch for sender during transaction introspection
  - Transaction simulation RPC call

5 network calls to Solana RPC during settlement
- **Settle Endpoint**:
  - Calls verify again = 4 calls
  - Transaction broadcast to network

Small note here but may be a larger discussion. Is it necessary to have verify AND settle?
It seems faster and more stable for the flow to be one network call from the server to the facilitator to verify and settle.
That's what the current settle endpoint does anyway.

### Network ID Handling Considerations

Solana "network IDs" (devnet = 103, testnet = 102, mainnet = 101) may overlap with EVM network IDs.

**Open Questions**:
- Should we use [CAIP-2](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-2.md) for networks?
- CAIP-2 + CAIP-19 for assets?
- Alternative: Create our own scheme (e.g. NetworkEnum, NetworkEnum:asset) and base everything off of it

**Some Affected Components**:
- `getDefaultAsset` in `x402/src/shared/middleware.ts`
- `getUsdcAddressForChain` in `x402/src/shared/evm/usdc.ts`
- `selectPaymentRequirements` in `x402/src/client/selectPaymentRequirements.ts`
- `getNetworkId` in `x402/src/shared/network.ts`

### Transaction Introspection

**Current Support**: Only TransferChecked from token and token-2022 programs
**Open Question**: Do we want to support Transfer as well?

### Examples Structure

Created two different versions of extending the examples:

1. **New Directory Approach** (`examples/typescript/facilitators/svm/express`)
   - More overhead but can be more customizable
   - Allows for SVM-specific configurations and dependencies
   - Cleaner separation of concerns

2. **File Addition Approach** (`examples/typescript/clients/fetch/solana.ts`)
   - Less overhead and complexity
   - Requires importing both EVM and Solana dependencies into the same example directory
   - More compact structure

**Open Question**: which approach is preferred?

## Testing
- [x] All existing tests pass
- [x] New SVM-specific tests added and passing
- [x] Error handling and edge case coverage

## Documentation
- Added comprehensive SVM scheme specification in `specs/schemes/exact/scheme_exact_svm.md`
- Updated examples and their READMEs

## Remaining Items (After receiving feedback on this PR)

### Package Support
- [ ] x402-axios
- [ ] x402-hono  
- [ ] x402-next

### Examples (Pending Structure Decision)
- [ ] Clients
  - [ ] axios
- [ ] Servers
  - [ ] express
  - [ ] hono
  - [ ] mainnet
- [ ] Example of how to set up an ATA (associated token account) for a merchant/resource server, unique to (asset, wallet)

### Other Features
- [ ] Paywall HTML implementation
- [ ] Squads and other major programmable wallet support
- [ ] Nonce-based transaction support?
- [ ] Token and Token-2022 Transfer instruction support (in addition to TransferChecked)

## Next Steps
- [ ] Gather feedback on example structure approach (separate directories vs. file addition)
- [ ] Decide on fee payer endpoint architecture.
- [ ] Determine network ID handling strategy
- [ ] Implement Remaining Items


## Tests

<!--
Please describe the tests you've performed to verify your changes.
Include relevant code samples, unit test cases, or screenshots if applicable.

For TypeScript: Run `pnpm test` from the `/typescript` directory
For Go: Run `go test ./...` from the `/go` directory
-->

## Checklist

- [x] I have formatted and linted my code
- [x] All new and existing tests pass
- [x] My commits are signed (required for merge) 

---

Thank you for reading!  

It's been exciting working on this and I'm happy to adjust it as needed to fit the community's shared vision.

cb-heimdall commented:
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td><code>2</code> if repo is sensitive</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>
<!-- CODEOWNERS & Bar Raisers -->
<!-- Power Rangers -->
<!-- Links -->

ngundotra commented on typescript/packages/x402/src/schemes/exact/svm/client.ts:
```diff
@@ -0,0 +1,213 @@
+  client: KeyPairSigner,
+  paymentRequirements: PaymentRequirements,
+) {
+  const transferIx = await createTransferInstruction(client, paymentRequirements);
+  const rpc = getRpcClient(paymentRequirements.network);
+  const { value: latestBlockhash } = await rpc.getLatestBlockhash().send();
+  const feePayer = paymentRequirements.extra?.feePayer as Address;
+
+  return pipe(
+    createTransactionMessage({ version: 0 }),
+    tx => setTransactionMessageFeePayer(feePayer, tx),
+    tx => setTransactionMessageLifetimeUsingBlockhash(latestBlockhash, tx),
+    tx => appendTransactionMessageInstructions([transferIx], tx),
```
This look awesome, can you also add support for ComputeBudget + PriorityFee, and ATA creation for the destination address?

In my notes, I had an "ideal" template transaction of following form:
```
[
  ComputeBudget.setLimit, 
  ComputeBudget.setPriorityFee, 
  ATA.createTokenAccountIdempotent, 
  TokenProgram*.transferChecked({account}, to, {mint}, decimals)
]
feePayer=facilitator
signers=[facilitator, {account}]
```

In this case `facilitator` =`feePayer`.

ngundotra commented on typescript/packages/x402/src/schemes/exact/svm/facilitator/verify.ts:
```diff
@@ -0,0 +1,261 @@
+    rpc,
+  );
+
+  // verify that the transaction and the transfer instruction are valid
```
Can we verify a slightly more complicated scheme here?

```
[
  ComputeBudget.setLimit, 
  ComputeBudget.setPriorityFee, 
  ATA.createTokenAccountIdempotent, 
  TokenProgram*.transferChecked({account}, to, {mint}, decimals)
]
feePayer=facilitator
signers=[facilitator, {account}]
```

I think it's pretty important to have priority fee, compute budget, and ATA creation in the transaction as well.

notorious-d-e-v commented:
Thank you for the review @ngundotra! Will be working these shortly.

notorious-d-e-v commented on typescript/packages/x402/src/types/shared/config.ts:
```diff
@@ -21,9 +22,17 @@ export const config: Record<string, ChainConfig> = {
     usdcName: "Bridged USDC",
   },
+  "solana-devnet": {
+    usdcAddress: "Gh9ZwEmdLJ8DscKNTkTqPbNwLNNBjuSzaG9Vp2KGtKJr" as SolanaAddress,
```
Good catch thanks!

notorious-d-e-v commented on typescript/packages/x402/src/shared/middleware.ts:
```diff
@@ -89,6 +91,22 @@ export function findMatchingRoute(
+  // solana devnet
+  if (network === NetworkEnum.SOLANA_DEVNET) {
+    return {
+      address: "Gh9ZwEmdLJ8DscKNTkTqPbNwLNNBjuSzaG9Vp2KGtKJr",
```
Good catch thanks!

notorious-d-e-v commented on typescript/packages/x402/src/schemes/exact/svm/client.ts:
```diff
@@ -0,0 +1,213 @@
+  const { asset } = paymentRequirements;
+
+  const rpc = getRpcClient(paymentRequirements.network);
+  const tokenAccount = await rpc.getAccountInfo(asset as Address).send();
```
Huge savings!

ngundotra commented:
@erikreppel-cb could we get a cursory review of the structure of this PR? It's 77 files, so want to get your feedback early!

CarsonRoscoe reviewed (commented):
Thank you for putting together this PR, and the foundation for helping out. This is looking promising to me.

I'm going to start going over the code, but first here are some high-level topics. 

### Fee Payer

I appreciate you calling this out, and you putting it in the extra field of payment requirements. That is appropriate. 

I am concerned about the need to add a new facilitator api endpoint to support this. This is a smell to me that future expansions will require similar custom work. I think we can come up with a more framework agnostic solution.

My proposal would be to enhance the [supported kinds response type](https://github.com/coinbase/x402/blob/main/typescript/packages/x402/src/types/verify/x402Specs.ts#L106) (e.g. [here](https://www.x402.org/facilitator/supported)) such that it can provide extra metadata around a specific network/scheme implementation. Essentially adding an `extra` or `metadata` field to `SupportedPaymentKindSchema` that the fee payer could go into. This achieves the same goal of allowing a different fee payer based on scheme and network if necessary.

You likely missed this since it isn't currently being used by the middleware, but we should add this API call to `useFacilitator`, and leverage it to discover the fee payer for a given network/scheme combination when checking what network/schemes the facilitator supports. 

cc: @erikreppel-cb thoughts?

###  "Is it necessary to have verify AND settle"

The short answer is yes. 

The long answer is that if we just settle before the server does work, the client will be wrongfully charged if the server throws an error during work. However, if we settle after doing the work and the payment fails, the server may have spent real money doing work it did not get paid for. Sandwiching the servers work between /verify and /settle in the middleware ensures that both parties are treated fairly. Servers don't work for free, but clients don't get charged for work that fails.

Mind you, this is just for our middleware implementations. Builders can always implement their own middleware that skips validation, or settles ahead of time, if needed. But our middleware should be robust and safe

### Network IDs

We use the term network ids over chain ids to avoid collisions between protocol families such as EVM and SVM. This is why we call it `base` and `base-sepolia` over listing the chain ids. In the evm implementation of the exact schema, we convert from network id to evm chain id once we actually need to use the chain id.

I suggest Solana does the same, and follows the same naming as cdp-sdk use for their network ids. That is, `solana`, `solana-devnet` and `solana-testnet` would be the network id names I'd suggest. 

As for CAIP-2/ CAIP-2 + CAIP-19, that is an interesting alternative. That may be the right long term approach, but it would affect the current EVM implementation. I will discuss this idea with the team and get back to you with our thoughts.

### Examples Structure

Personally, I lean towards the file addition approach. I think it maps more nicely to our packages. You see `x402-axios`, you go to `examples/typescript/clients/axios`, and you see the EVM and SVM examples for that package. I think that is easier to navigate for builders reading the repo

ngundotra commented on typescript/packages/x402/src/schemes/exact/svm/client.ts:
```diff
@@ -0,0 +1,423 @@
+  instructions: Instruction[],
+): Promise<number> {
+  // get the addresses of the write locked accounts
+  const writeLockedAccounts: Address[] = [
+    ...new Set(
+      instructions.flatMap(
+        instruction =>
+          instruction.accounts
+            ?.filter(account => account.role === AccountRole.WRITABLE)
+            .map(account => account.address) ?? [],
+      ),
+    ),
+  ];
+
+  // call RPC to get the recent prioritization fees on the write-locked accounts
+  const recentPrices = await rpc.getRecentPrioritizationFees(writeLockedAccounts).send();
+
+  if (recentPrices.length === 0) {
+    return 10_000_000; // default to 10 lamports, same as phantom wallet
+  }
+
+  // take the 90th percentile of the recent prices
+  const sortedPrices = [...recentPrices].sort((a, b) =>
+    Number(a.prioritizationFee - b.prioritizationFee),
+  );
+  const percentileIndex = Math.floor(0.9 * (sortedPrices.length - 1));
+  const percentilePrice = sortedPrices[percentileIndex].prioritizationFee;
+  const bidPrice = Math.ceil(Number(percentilePrice));
+
+  if (bidPrice === 0) {
+    return 10_000_000; // default to 10,000,000 microlamports, same as phantom wallet
+  }
```
This should be set to a default of 1 microlamport to reduce cost on facilitator side. 

I think having a dynamic methodology should come later for high throughput facilitators that are processing large # of transactions. Those facilitators have an explicit interest in guaranteeing the long tail of transfers land. These apps (like Phantom) are very sensitive to even 0.1% of transfers failing, as it results in worse UX.

ngundotra commented on typescript/packages/x402/src/schemes/exact/svm/client.ts:
```diff
@@ -0,0 +1,423 @@
+): Promise<TransferCheckedInstructionToken2022> {
+  const { asset, maxAmountRequired: amount, payTo } = paymentRequirements;
+
+  const [sourceATA] = await findAssociatedTokenPdaToken2022({
+    mint: asset as Address,
+    owner: client.address,
+    tokenProgram: TOKEN_2022_PROGRAM_ADDRESS,
+  });
+
+  const [destinationATA] = await findAssociatedTokenPdaToken2022({
+    mint: asset as Address,
+    owner: payTo as Address,
+    tokenProgram: TOKEN_2022_PROGRAM_ADDRESS,
+  });
+
+  return getTransferCheckedInstructionToken2022(
+    {
+      source: sourceATA,
+      mint: asset as Address,
+      destination: destinationATA,
+      authority: client,
+      amount: BigInt(amount),
+      decimals: decimals,
+    },
+    { programAddress: TOKEN_2022_PROGRAM_ADDRESS },
+  );
+}
+
+/**
+ * Creates an spl-token transfer instruction for the given client and payment requirements.
+ * This function will create a transfer instruction for a token created
+ * by the token program: https://github.com/solana-program/token/
+ *
+ * @param client - The signer instance who's tokens will be debited from
+ * @param paymentRequirements - The payment requirements
+ * @param decimals - The decimals of the token
+ * @returns A promise that resolves to the transfer instruction
+ */
+async function createTransferInstructionToken(
+  client: KeyPairSigner,
```
These two methods can be one function, can just take a single token program as argument

ngundotra commented on typescript/packages/x402/src/schemes/exact/svm/facilitator/settle.ts:
```diff
@@ -0,0 +1,231 @@
+    };
+
+    // create the config for the transaction confirmation
+    const commitment: Commitment = "finalized";
```
I believe it is better to set `confirmed`, confirmations don't roll back.

notorious-d-e-v commented on typescript/packages/x402/src/schemes/exact/svm/facilitator/settle.ts:
```diff
@@ -0,0 +1,231 @@
+    assertIsTransactionMessageWithBlockhashLifetime(decompiledTransactionMessage);
+
+    // add the blockhash lifetime to the signed transaction
+    const signedTransactionWithBlockhashLifetime = {
+      ...signedTransaction,
+      lifetimeConstraint: decompiledTransactionMessage.lifetimeConstraint,
+    };
```
I think this was related to typescript type hell with solana kit. 

Maybe there's another way to do it like passing it through a solana kit assert function that I don't know about.

notorious-d-e-v commented:
@CarsonRoscoe I've went ahead and replaced the `/fee-payer` route with an `extra` field returned in the `PaymentKindsResponse` from the facilitator in the `/supported` route.

Updated the `useFacilitator` to support the `/supported` route.

I've also adjusted the examples to be files instead of directories.

The network names are now `solana` and `solana-devnet` and the network ids reflect those.

Please let me know if there's anything else.

CarsonRoscoe reviewed (commented):
Completed the initial manual testing, and I am very happy! The implementation is looking promising and well-structured. 

<img width="784" height="189" alt="Screenshot 2025-08-06 at 12 11 04 PM" src="https://github.com/user-attachments/assets/4560043d-0e07-4aca-bf1a-5cc61e16b1d6" />

Below are the key feedback points and requested changes.

## Requested Changes

### 1. Example Structure Unification

#### Server Examples
- Server examples should be unified between Ethereum and Solana implementations
  - Server code is agnostic to the chain - it only deals with public addresses/keys and network definitions

#### Client Examples
- Client examples can remain separate, but I would prefer to unify them between Ethereum and Solana as well. If we can get the branching wallet logic to be very minimal, and showcase the example supporting either a evm or svm wallet via the same sdk interface, that would be ideal.
  - Client code is not agnostic to the chain - but it does use the same sdk entrypoint.

#### Facilitator Example
- Facilitator example should also be merged into one. If we can have this be one example that takes a EVM_PRIVATE_KEY and a SVM_PRIVATE_KEY, and supports base-sepolia as well as solana-devnet, that would be ideal

### 2. Signer Implementation

#### Package Level Changes
- Update `createSignerFromBase58` to support multiple private key formats in signer creation:
  - Option 1: Decoded private key
  - Option 2: Byte array input
  - Both should resolve to the same signer type

#### Example Standardization
- Currently `createSignerFromBase58` exists in:
  - Client example
  - Shared SVM folder
- Update all examples to use helpers exported from middleware
- Keep examples minimal and consistent

### 3. Documentation Updates

#### README Consolidation
- Combine separate example READMEs into core README
- Facilitator README should cover both EVM and Solana implementations
- Emphasize unified system approach:
  - Facilitators can support both EVM and SVM
  - Middleware can handle payments from either chain
  - Example: Endpoint accepting $0.10 USDC on either Base or Solana
- Avoid suggesting EVM and SVM as separate development paths

### 4. Testing

#### E2E Testing
- Currently missing e2e tests for Solana, however the e2e testing package also lacks support for different protocol implementations.
- Action Item (@CarsonRoscoe ):
  - Will create PR to add Solana e2e test support
  - Will implement express/fetch test cases
- Action Item (@notorious-d-e-v ):
  - Implement e2e tests for remaining middleware and clients after support is added

## Overall Assessment

The implementation is solid and working well in practice! These requested changes are mostly around examples, testing, and minimizing duplicate code.

Thank you for all your work, this is looking fantastic, and I'm excited to get this in :)

CarsonRoscoe reviewed (commented):
Actually, am I correct in understanding we've lost `payer` in settlement response, as a result of `ExactSvmPayloadSchema` not having the equivalent `authorization.from` value to read from?

Is it possible to get the `payer` our of the `transaction` in `ExactSvmPayloadSchema`? If not, is it possible to include a payer field in `ExactSvmPayloadSchema` that the facilitator can validate is who that transaction belongs to? It would just be ideal to have the facilitator return `payer` to the resource server, as the server may choose to create a side-effect regarding it (i.e., in their own internal database, track the successful sale with that payer).

notorious-d-e-v commented:
Thank you for the review and feedback @CarsonRoscoe!

Question regarding the Action Items for e2e testing: do we want to get those done in this PR or in a follow-on?

Everything else is clear - I'll go ahead and get those changes in.

> Actually, am I correct in understanding we've lost `payer` in settlement response, as a result of `ExactSvmPayloadSchema` not having the equivalent `authorization.from` value to read from?
> 
> Is it possible to get the `payer` out of the `transaction` in `ExactSvmPayloadSchema`? If not, is it possible to include a payer field in `ExactSvmPayloadSchema` that the facilitator can validate is who that transaction belongs to? It would just be ideal to have the facilitator return `payer` to the resource server, as the server may choose to create a side-effect regarding it (i.e., in their own internal database, track the successful sale with that payer).

Regarding the payer in settlement response: as you mentioned, the payer can be retrieved from the transaction. 
We do currently retrieve and return it to the server if verification/settlement is _successful_.

However, if verification fails before we get to checking the transfer instruction in the transaction, then `payer` is returned as an empty string. This could happen due to the following reasons during verification:
	1. the `PaymentPayload` (network or scheme) does not match the `PaymentRequirement` (network or scheme), or the `PaymentPayload` network is not a supported svm network.
	2. the transaction sent by the client does not match the transaction that is expected by the server/facilitator. For example, the number of instructions are different, or the required instructions are not present.
	3. the receiver address (ATA) does not exist and needs to be created, but there is no instruction included in the transaction to create it.

We could add some logic to get the `payer` earlier on in the process, so that it is always available in the verification response, barring some failures like a malformed transaction, or a transfer instruction not existing inside the transaction.

We can also provide it as part of the `ExactSvmPayloadScheme`. It would have to be signed by the client's private key, and the facilitator will have to do some extra work to check that the signature resolves to the payer, and that the payer matches the sender address in the transfer instruction of the transaction.

Currently payer is marked as optional in the schema, which is why we designed it to be returned later on in the verification process. 
Should we update the docs / schema, if it is required?

CarsonRoscoe reviewed (commented):
> Question regarding the Action Items for e2e testing: do we want to get those done in this PR or in a follow-on?

I suppose it can be a follow up. I'd like to get this merged in asap at this point.
As long as the E2E tests continue passing for the EVM implementation, we can merge it in and fast-follow.

> Regarding the payer in settlement response: as you mentioned, the payer can be retrieved from the transaction.
We do currently retrieve and return it to the server if verification/settlement is successful.

Noted, thank you for the clarification, I must have missed that

> The reason I added it to the client example was so that the client doesn't have to install the x402 package in addition to the x402 fetch package.

Yes, I would re-export, similar to how the x402 axios package currently re-exports the decodeXPaymentResponse function:
e.g. `export { decodeXPaymentResponse } from "x402/shared";` but for your use case

notorious-d-e-v commented:
I'm actually thinking of making the `MultiNetworkSigner` be an object mapping `{Network: Signer}` instead of `{evm: Signer, svm: Signer}` what do you think @CarsonRoscoe ? 

It will be useful especially for setting variable RPC's per network.

xR0am commented:
This if really great work @notorious-d-e-v !

Cant wait to see this merged so I can enhance my MCP.

https://github.com/xR0am/tip-md-x402-mcp-server

CarsonRoscoe reviewed (commented):
Hey @notorious-d-e-v,

In the current approach, there is no way for a Solana signer to be used for a Ethereum network, or vice-versa. There's an extra guard inherited from this explicit design. Also, requiring it's set per-network would be a degradation in devx for circumstances where  one wallet instance can satisfy many chains.

I do see the benefit of allowing a custom RPC to be used for specific chains, and thus the need to assign signers per chains. However, I don't think this should replace what we have, it should be in addition to. Perhaps something like:
```ts
type MultiNetworkSigner = { evm: evm.EvmSigner; svm: svm.SvmSigner } & Partial<Record<Network, Signer>>
```
where you CAN assign a signer per network as an override, while still assigning the default for a given family.

Regardless, I'd like to do that as a fast-follow. It would be non-breaking to upgrade the interface to that if we do it right, so I am comfortable prioritizing getting this in.

Just completing a final round of testing!

CarsonRoscoe reviewed (approved):
Thank you so much for the contribution, fantastic work!

Merging this in! I'll fast-follow with the e2e testing (#352 should unblock), and we should potentially also follow with the network/signer changes discussed as well

ukstv commented:
Yahoo!!!

notorious-d-e-v commented:
> Thank you so much for the contribution, fantastic work!
> 
> Merging this in! I'll fast-follow with the e2e testing (#352 should unblock), adding solana-devnet to the x402.org default facilitator, and we should potentially also follow with the network/signer changes discussed as well

Woohoo! Sounds good! Big thanks to you and @ngundotra for design and feedback!

Abelhubprog commented:
Huge, congratulations 👏

notorious-d-e-v commented:
> This if really great work @notorious-d-e-v !
> 
> Cant wait to see this merged so I can enhance my MCP.
> 
> https://github.com/xR0am/tip-md-x402-mcp-server

Looking forward to it!
